File: gyroball/App/SessionTracker.swift
import Foundation
import Combine

/// Watches the live BLE stream and turns it into persisted workout sessions.
/// A session opens on the first packet and closes after `gracePeriod` without
/// packets, so the ball dropping BLE below ~1700 RPM and being spun back up
/// shortly after stays within one session.
final class SessionTracker {

    private let ble: BLEManager
    private let store: SessionStore

    private var sessionStart: Date?
    private var samples: [Double] = []
    private var lastSampleAt: Date?
    private var endTimer: Timer?
    private var cancellables: Set<AnyCancellable> = []

    private let gracePeriod: TimeInterval = 45
    private let minimumRevolutions: Double = 30

    init(ble: BLEManager, store: SessionStore) {
        self.ble = ble
        self.store = store

        ble.packetTick
            .receive(on: DispatchQueue.main)
            .sink { [weak self] rpm in self?.handleTick(rpm) }
            .store(in: &cancellables)
    }

    private func handleTick(_ rpm: Double) {
        let now = Date()

        if sessionStart == nil {
            sessionStart = now
            samples = []
            lastSampleAt = nil
        }

        if lastSampleAt == nil || now.timeIntervalSince(lastSampleAt!) >= 1.0 {
            samples.append(rpm)
            lastSampleAt = now
        }

        endTimer?.invalidate()
        endTimer = .scheduledTimer(withTimeInterval: gracePeriod, repeats: false) { [weak self] _ in
            self?.finalize()
        }
    }

    /// Persists the current session (if it was a real workout) and resets.
    func finalize() {
        defer { discard() }
        guard let start = sessionStart else { return }

        let t = ble.telemetry
        guard t.totalRevolutions >= minimumRevolutions else { return }

        store.add(WorkoutSession(
            startedAt: start,
            duration: t.activeSeconds,
            topRPM: t.topSpeed,
            avgRPM: t.averageRPM,
            revolutions: t.totalRevolutions,
            samples: samples))
    }

    /// Throws away the current session without saving.
    func discard() {
        endTimer?.invalidate()
        endTimer = nil
        sessionStart = nil
        samples = []
        lastSampleAt = nil
        ble.resetSession()
    }
}

File: windows/src/Gyroball.Core/SessionTracker.cs
using Gyroball.Core.Storage;

namespace Gyroball.Core;

/// <summary>
/// Turns the live packet stream into persisted sessions. A session opens on the
/// first packet and closes after <see cref="GracePeriod"/> without packets, so
/// the ball dropping BLE below ~1700 RPM and being spun back up shortly after
/// stays within one session. Ported from SessionTracker.swift.
///
/// Grace expiry is driven by <see cref="CheckGrace"/>, which the host calls on a
/// periodic timer; tests call it directly with a chosen "now".
/// </summary>
public sealed class SessionTracker
{
    public const double GracePeriod = 45;          // seconds of silence before a session closes
    public const double MinimumRevolutions = 30;   // ignore trivial spins

    private readonly LiveTelemetry _live;
    private readonly SessionStore _store;

    private DateTime? _sessionStart;
    private DateTime? _lastSampleAt;
    private DateTime? _lastTickAt;
    private readonly List<double> _samples = new();

    public SessionTracker(LiveTelemetry live, SessionStore store)
    {
        _live = live;
        _store = store;
        _live.PacketTick += HandleTick;   // (rpm, timestamp)
    }

    public void HandleTick(double rpm, DateTime now)
    {
        if (_sessionStart is null)
        {
            _sessionStart = now;
            _samples.Clear();
            _lastSampleAt = null;
        }

        if (_lastSampleAt is null || (now - _lastSampleAt.Value).TotalSeconds >= 1.0)
        {
            _samples.Add(rpm);
            _lastSampleAt = now;
        }

        _lastTickAt = now;
    }

    /// <summary>Closes the session if the grace period has elapsed since the last packet.</summary>
    public void CheckGrace(DateTime now)
    {
        if (_sessionStart is null || _lastTickAt is null) return;
        if ((now - _lastTickAt.Value).TotalSeconds >= GracePeriod) FinalizeSession();
    }

    /// <summary>Persists the current session (if it was a real workout) and resets.</summary>
    public void FinalizeSession()
    {
        try
        {
            if (_sessionStart is not { } start) return;
            if (_live.TotalRevolutions < MinimumRevolutions) return;

            _store.Add(new WorkoutSession
            {
                StartedAt = start,
                Duration = _live.ActiveSeconds,
                TopRpm = _live.TopSpeed,
                AvgRpm = _live.AverageRpm,
                Revolutions = _live.TotalRevolutions,
                Samples = _samples.ToArray()
            });
        }
        finally
        {
            Discard();
        }
    }

    /// <summary>Throws away the current session without saving.</summary>
    public void Discard()
    {
        _sessionStart = null;
        _lastSampleAt = null;
        _lastTickAt = null;
        _samples.Clear();
        _live.ResetSession();
    }
}
